src/Prove.hs

```diff
@@ -0,0 +1,249 @@
+            postVars = Map.fromList $ catInts poststate
+            lhs = fromMaybe
+                    (error (show item <> " not found in " <> show pre))
+                    $ Map.lookup (nameFromItem pre item) preVars
+        rhs <- symExpInt post e
+        return $ lhs .== rhs
+
+
+
+updates :: Map Id [Either StorageLocation StorageUpdate] -> [(Id, StorageUpdate)]
+-- TODO: handle storage reads as well as writes
+updates stateUpdates = mkPairs $ fmap rights stateUpdates
+
+mkPairs :: Map Id [StorageUpdate] -> [(Id, StorageUpdate)]
+mkPairs updates = concat $ fmap (\(c, us) -> fmap (\u -> (c, u)) us) (Map.toList updates)
+
+makeSymbolic :: Ctx -> StorageLocation -> Symbolic (Id, SMType)
+makeSymbolic ctx loc = case loc of
+    IntLoc item -> do
+      let name = nameFromItem ctx item
+      v <- sInteger name
+      return $ (name, SymInteger $ v)
+    BoolLoc item -> do
+      let name = nameFromItem ctx item
+      v <- sBool name
+      return $ (name, SymBool $ v)
+    l -> error ("TODO: handle " ++ show l ++ " in makeSymbolic")
+
+symExp :: Ctx -> ReturnExp -> Symbolic (SMType)
+symExp ctx expression = case expression of
+  ExpInt e -> do
+    i <- symExpInt ctx e
+    return $ SymInteger i
+  ExpBool e -> do
+    b <- symExpBool ctx e
+    return $ SymBool b
+  ExpBytes e -> do
+    b <- symExpBytes ctx e
+    return $ SymBytes b
+
+symExpBool :: Ctx -> Exp Bool -> Symbolic (SBV Bool)
+symExpBool ctx@(Ctx _ _ _ store) e = case e of
+  And a b   -> (.&&) <$> (symExpBool ctx a) <*> (symExpBool ctx b)
+  Or a b    -> (.||) <$> (symExpBool ctx a) <*> (symExpBool ctx b)
+  Impl a b  -> (.=>) <$> (symExpBool ctx a) <*> (symExpBool ctx b)
+  Eq a b    -> (.==) <$> (symExpInt ctx a) <*> (symExpInt ctx b)
+  LE a b    -> (.<)  <$> (symExpInt ctx a) <*> (symExpInt ctx a)
+  LEQ a b   -> (.<=) <$> (symExpInt ctx a) <*> (symExpInt ctx a)
+  GE a b    -> (.>)  <$> (symExpInt ctx a) <*> (symExpInt ctx a)
+  GEQ a b   -> (.>=) <$> (symExpInt ctx a) <*> (symExpInt ctx a)
+  NEq a b   -> sNot  <$> (symExpBool ctx (Eq a b))
+  Neg a     -> sNot  <$> (symExpBool ctx a)
+  LitBool a -> return $ literal a
+  BoolVar a -> sBool a -- TODO: handle calldata args properly
+  TEntry a  -> do
+    let vars = Map.fromList $ catBools store
+    return
+      $ fromMaybe (error (show a <> " not found in " <> show vars))
+      $ Map.lookup (nameFromItem ctx a) vars
+
+symExpInt :: Ctx -> Exp Integer -> Symbolic (SBV Integer)
+symExpInt ctx@(Ctx _ _ _ store) e = case e of
+  Add a b   -> (+)  <$> (symExpInt ctx a) <*> (symExpInt ctx b)
+  Sub a b   -> (-)  <$> (symExpInt ctx a) <*> (symExpInt ctx b)
+  Mul a b   -> (*)  <$> (symExpInt ctx a) <*> (symExpInt ctx b)
+  Div a b   -> sDiv <$> (symExpInt ctx a) <*> (symExpInt ctx b)
+  Mod a b   -> sMod <$> (symExpInt ctx a) <*> (symExpInt ctx b)
+  Exp a b   -> (.^) <$> (symExpInt ctx a) <*> (symExpInt ctx b)
+  LitInt a  -> return $ literal a
+  IntVar a  -> sInteger a -- TODO: handle calldata args properly
```

Comment:
It seems that if we keep calldata args in `Ctx` as well, this method (as well as `symExpBool`, `symExpBytes` and `symExp`) could be made pure and things would probably look a little cleaner.

Reply:
good point! it's done.